Export role details in JSON format

slatinsky opened the pull request:
<!--

**Important:**

Please make sure that there is an existing issue that describes the problem solved by your pull request. If there isn't one, consider creating it first.
An open issue offers a good place to iron out requirements, discuss possible solutions, and ask questions.

Remember to also:

- Keep your pull request focused and as small as possible. If you want to contribute multiple unrelated changes, please create separate pull requests for each of them.
- Follow the coding style and conventions already established by the project. When in doubt about which style to use, ask in the comments to your pull request.
- If you want to start a discussion regarding a specific change you've made, add a review comment to your own code. This can be used to highlight something important or to seek further input from others.

-->

<!-- Please specify the issue addressed by this pull request -->


This pull request adds `roles` field to `messages.author` in JSON output format.

If no roles are available, empty array is created:
```json
"author": {
	...
	"roles": [],
	...
},
```
If role color is not available, `null` value is used. Else the color is is converted to hexcode

```json
"roles": [
	{
		"id": "712543702037017113",
		"name": "mod",
		"color": null,
		"position": 239
	},
	{
		"id": "766532373335259678",
		"name": "Orange",
		"color": "#F58231",
		"position": 218
	},
	...
]
```

Roles are sorted by position, as they are in the Discord client

Tested CLI export in Json format - DM channel (no roles) and Guild channel (with roles). \
GUI not tested, but the changes should not affect it.

Tyrrrz commented on DiscordChatExporter.Core/Exporting/ExportContext.cs:
```diff
@@ -106,6 +106,18 @@ await foreach (var role in Discord.GetGuildRolesAsync(Request.Guild.Id, cancella
             .FirstOrDefault();
     }
 
+    public IEnumerable<Role> TryGetUserRoles(Snowflake id)
```
Let's make this `IReadOnlyList<Role> GetUserRoles(...)` (no need for `Try...` because it doesn't return `null`).

Tyrrrz replied:
Also, let's update `TryGetUserColor(...)` so that it calls `GetUserRoles(...)` for code reuse

slatinsky commented:
In the last commit:
- Renamed `IEnumerable<Role> TryGetUserRoles(...)` to `IReadOnlyList<Role> GetUserRoles(...)`
- Simplified `TryGetUserColor(...)`
- Tested build on Windows + CLI channel export in Json format

Tyrrrz commented:
Thanks!

stepech commented:
Hey, thanks for change.
Was there some consideration about how large the file will get with this change? Is it recorded really for each message including on servers with 100s of roles? What about creating extra file during export, which would hold just dictionary of users and their corresponding roles?

slatinsky commented:
Discord limits roles to 250 per server. That is the edge case scenario, but many users in my experience have only small number of roles. Yes, it is recorded in every message.
You could deduplicate a lot of things if you really wanted - reactions, authors, mentions, stickers. This felt like the best place to put it considering the current state of the project.
Yes, it will increase the file size. But text size increase is minimal in my opinion vs. the size of assets. And hard drive space is very cheap. And we were fetching this information anyway, just not recording it in the file.
